File: backend/app/schemas/stock.py
"""
Pydantic v2 schemas — Stocks, Quotes, Historical Prices, Company Profiles.
"""

from datetime import date, datetime

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Company Profile
# ---------------------------------------------------------------------------

class CompanyProfileSchema(BaseModel):
    ticker: str
    company_name: str
    exchange: str | None
    sector: str | None
    industry: str | None
    country: str | None
    currency: str | None
    website: str | None
    logo_url: str | None
    description: str | None
    ceo: str | None
    employees: int | None
    ipo_date: date | None
    market_cap: float | None
    price: float | None
    beta: float | None
    is_etf: bool
    is_actively_trading: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Stock Quote
# ---------------------------------------------------------------------------

class StockQuoteSchema(BaseModel):
    ticker: str
    price: float | None
    change: float | None
    change_percent: float | None
    volume: int | None
    avg_volume: int | None
    open: float | None
    high: float | None
    low: float | None
    previous_close: float | None
    high_52w: float | None
    low_52w: float | None
    market_cap: float | None
    pe_ratio: float | None
    eps: float | None
    timestamp: datetime | None
    is_market_open: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Historical Price
# ---------------------------------------------------------------------------

class HistoricalPriceSchema(BaseModel):
    date: date
    open: float | None
    high: float | None
    low: float | None
    close: float | None
    adj_close: float | None
    volume: int | None
    change: float | None
    change_percent: float | None

    model_config = {"from_attributes": True}


class HistoricalPriceResponse(BaseModel):
    ticker: str
    data: list[HistoricalPriceSchema]


# ---------------------------------------------------------------------------
# Search
# ---------------------------------------------------------------------------

class StockSearchResult(BaseModel):
    ticker: str
    name: str
    exchange: str | None
    sector: str | None
    currency: str | None
    logo_url: str | None


class StockSearchResponse(BaseModel):
    query: str
    results: list[StockSearchResult]
    total: int


# ---------------------------------------------------------------------------
# Stock Detail (aggregated)
# ---------------------------------------------------------------------------

class StockDetailResponse(BaseModel):
    profile: CompanyProfileSchema
    quote: StockQuoteSchema | None


# ---------------------------------------------------------------------------
# Pagination
# ---------------------------------------------------------------------------

class PaginationMeta(BaseModel):
    page: int
    page_size: int
    total: int
    total_pages: int

File: frontend/src/types/stock.ts
/**
 * Stock & market data types.
 * Matches backend/app/schemas/stock.py (to be kept in sync).
 */

export interface CompanyProfile {
  id: string;
  ticker: string;
  name: string;
  exchange: string | null;
  sector: string | null;
  industry: string | null;
  description: string | null;
  website: string | null;
  logo_url: string | null;
  ceo: string | null;
  employees: number | null;
  country: string | null;
  currency: string | null;
  market_cap: number | null;
  ipo_date: string | null;
}

export interface StockQuote {
  ticker: string;
  price: number | null;
  change: number | null;
  change_percent: number | null;
  volume: number | null;
  avg_volume: number | null;
  market_cap: number | null;
  pe_ratio: number | null;
  eps: number | null;
  high_52w: number | null;
  low_52w: number | null;
  open: number | null;
  previous_close: number | null;
  timestamp: string | null;
}

export interface HistoricalPrice {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  adj_close: number | null;
  volume: number;
  vwap: number | null;
}

export interface StockDetail {
  profile: CompanyProfile;
  quote: StockQuote;
}

export interface SearchResult {
  ticker: string;
  name: string;
  exchange: string | null;
  sector: string | null;
}
